core/src/main/java/com/novoda/noplayer/internal/exoplayer/NoPlayerAdsLoader.java

```diff
@@ -154,25 +160,71 @@ public void onTimelineChanged(Timeline timeline, @Nullable Object manifest, int
         if (player != null) {
             adGroupIndex = player.getCurrentAdGroupIndex();
             adIndexInGroup = player.getCurrentAdIndexInAdGroup();
+
+            if (reason == Player.TIMELINE_CHANGE_REASON_PREPARED) {
+                notifyAdvertStart();
+            }
         }
     }
 
     @Override
     public void onPositionDiscontinuity(int reason) {
         if (reason == Player.DISCONTINUITY_REASON_AD_INSERTION && player != null && adPlaybackState != null) {
             if (adGroupIndex != -1 && adIndexInGroup != -1) {
+                notifyAdvertEnd();
                 adPlaybackState = adPlaybackState.withPlayedAd(adGroupIndex, adIndexInGroup);
                 updateAdPlaybackState();
             }
 
             adGroupIndex = player.getCurrentAdGroupIndex();
             adIndexInGroup = player.getCurrentAdIndexInAdGroup();
+            notifyAdvertStart();
         }
     }
 
-    private void notifyEventIfPossible(String event) {
-        if (advertListener.isPresent()) {
-            advertListener.get().onAdvertEvent(event);
+    private void notifyAdvertEnd() {
+        AdvertBreak advertBreak = advertBreaks.get(adGroupIndex);
+        List<Advert> adverts = advertBreak.adverts();
+        advertListener.onAdvertEnd(adverts.get(adIndexInGroup).advertId());
+
+        if (adIndexInGroup == adverts.size() - 1) {
+            advertListener.onAdvertBreakEnd(advertBreak.advertBreakId());
+        }
+    }
+
+    private void notifyAdvertStart() {
+        if (adGroupIndex != -1 && adIndexInGroup != -1) {
```

Comment:
for `notifyAdvertEnd()` this check is external: worth keeping this outside too?

Reply:
55a7212